File: Next-Up-Hardwood-Hoops-main/courthoops/simulation/engine/foul_engine.py
from typing import Callable, Optional, Dict

from courthoops.domain.game.events import Event
from courthoops.domain.game.possession import Possession
from courthoops.infra.utils.rng import IRNG
from courthoops.simulation.playbyplay.event_builder import EventBuilder
from courthoops.domain.player.entity import Player


class FoulEngine:
    """Handles foul determination and consequences."""

    def __init__(
        self,
        rng: IRNG,
        offense_strength_fn: Callable[[str], float],
        defense_strength_fn: Callable[[str], float],
        player_lookup: Optional[Dict[str, Player]] = None,
        event_builder: Optional[EventBuilder] = None,
    ):
        self.rng = rng
        self.offense_strength_fn = offense_strength_fn
        self.defense_strength_fn = defense_strength_fn
        self.player_lookup = player_lookup or {}
        self.event_builder = event_builder or EventBuilder()

    def maybe_commit_foul(
        self,
        possession: Possession,
        defender_id: str | None = None,
        shooter_id: str | None = None,
        play_type: str | None = None,
        team_fouls: Optional[Dict[str, int]] = None,
    ) -> Optional[Event]:
        """Stochastic foul check influenced by defensive strength and bonus."""
        defense_strength = self.defense_strength_fn(possession.defense_team_id)
        offense_strength = self.offense_strength_fn(possession.offense_team_id)
        foul_prob = 0.08 + (offense_strength - defense_strength) / 300.0
        defender = self.player_lookup.get(defender_id) if defender_id else None
        if defender:
            foul_prob += (100 - defender.attributes.defensive_iq) / 1000.0
            foul_prob += (100 - defender.attributes.decision_discipline) / 1200.0
        foul_prob = min(max(foul_prob, 0.02), 0.2)
        if self.rng.random() >= foul_prob:
            return None
        desc = f"{possession.defense_team_id} commits a foul."
        shooting_foul = False
        ft_attempts = 0
        team_fouls = team_fouls or {}
        bonus = team_fouls.get(possession.defense_team_id, 0) >= 7  # simple single-bonus heuristic
        if bonus and self.rng.random() < 0.6:
            shooting_foul = True
            ft_attempts = 2
        elif self.rng.random() < 0.55:  # lean toward shooting fouls in this lightweight model
            shooting_foul = True
            # Simple heuristic: post/drive -> 2, spot-up sometimes 3
            if play_type == "spot_up" and self.rng.random() < 0.25:
                ft_attempts = 3
            else:
                ft_attempts = 2
        and_one = False
        if shooting_foul and self.rng.random() < 0.2:
            # mark as potential and-one if caller's shot was made
            and_one = True
        payload = {
            "team_id": possession.defense_team_id,
            "player_id": defender_id,
            "period": possession.period,
            "type": "personal",
            "shooting_foul": shooting_foul,
            "ft_attempts": ft_attempts,
            "fouled_player_id": shooter_id,
            "and_one": and_one,
        }
        return self.event_builder.build("foul", desc, payload)

File: Scripts/Core/Simulation/Engines/FoulEngine.cs
using System;
using System.Collections.Generic;
using HardwoodHoops.Core.Domain.Games;
using HardwoodHoops.Core.Domain.Players;
using HardwoodHoops.Core.Simulation.PlayByPlay;
using HardwoodHoops.Core.Simulation.Rng;
using DomainGameEvent = HardwoodHoops.Core.Domain.Games.GameEvent;

namespace HardwoodHoops.Core.Simulation.Engines;

public sealed class FoulEngine
{
    private readonly IRng _rng;
    private readonly Func<string, double> _offenseStrength;
    private readonly Func<string, double> _defenseStrength;
    private readonly Dictionary<string, Player> _playerLookup;
    private readonly EventBuilder _eventBuilder;

    public FoulEngine(
        IRng rng,
        Func<string, double> offenseStrength,
        Func<string, double> defenseStrength,
        Dictionary<string, Player>? playerLookup = null,
        EventBuilder? eventBuilder = null)
    {
        _rng = rng;
        _offenseStrength = offenseStrength;
        _defenseStrength = defenseStrength;
        _playerLookup = playerLookup ?? new Dictionary<string, Player>();
        _eventBuilder = eventBuilder ?? new EventBuilder();
    }

    public DomainGameEvent? MaybeCommitFoul(
        string offenseTeamId,
        string defenseTeamId,
        int period,
        string? defenderId,
        string? shooterId,
        string? playType,
        Dictionary<string, int>? teamFouls)
    {
        var defenseStrength = _defenseStrength(defenseTeamId);
        var offenseStrength = _offenseStrength(offenseTeamId);
        var foulProb = 0.08 + (offenseStrength - defenseStrength) / 300.0;

        if (defenderId is not null && _playerLookup.TryGetValue(defenderId, out var defender))
        {
            foulProb += (100 - defender.Attributes.DefensiveIq) / 1000.0;
            foulProb += (100 - defender.Attributes.DecisionDiscipline) / 1200.0;
        }

        foulProb = Math.Clamp(foulProb, 0.02, 0.2);
        if (_rng.NextDouble() >= foulProb)
        {
            return null;
        }

        var shootingFoul = false;
        var ftAttempts = 0;
        var bonus = teamFouls is not null && teamFouls.TryGetValue(defenseTeamId, out var fouls) && fouls >= 7;
        if (bonus && _rng.NextDouble() < 0.6)
        {
            shootingFoul = true;
            ftAttempts = 2;
        }
        else if (_rng.NextDouble() < 0.55)
        {
            shootingFoul = true;
            if (playType == "spot_up" && _rng.NextDouble() < 0.25)
            {
                ftAttempts = 3;
            }
            else
            {
                ftAttempts = 2;
            }
        }

        var andOne = shootingFoul && _rng.NextDouble() < 0.2;
        var payload = new Dictionary<string, object>
        {
            ["team_id"] = defenseTeamId,
            ["player_id"] = defenderId ?? string.Empty,
            ["period"] = period,
            ["type"] = "personal",
            ["shooting_foul"] = shootingFoul,
            ["ft_attempts"] = ftAttempts,
            ["fouled_player_id"] = shooterId ?? string.Empty,
            ["and_one"] = andOne
        };

        return _eventBuilder.Build("foul", $"{defenseTeamId} commits a foul.", payload);
    }
}
